### PR TITLE
Update/harden docker.elastic.co/elasticsearch/elasticsearch (#5292)

### LINKED ISSUE #5292: Update/harden docker.elastic.co/elasticsearch/elasticsearch
We currently use

https://www.docker.elastic.co/r/elasticsearch/elasticsearch:7.10.1-amd64

from Elastic's registry which turns out to be the same as the one on docker.io:

https://hub.docker.com/layers/library/elasticsearch/7.10.1/images/sha256-e9a1fe65f68b2d2b9583287d1190f67f23af08582eac4d2a8dc342e4219c7306

Elastic offers an image for the next patch level, 7.10.2, and docker.io does not, but that version is only two months younger and so will be equally riddled with CVEs.

Worse, the images from that time period (2020/21) are based on CentOS 8 which is EOL, so without updates.

More recent 7.x images are based on Ubuntu LTS 20 which is not yet EOL. This means we need to update to the most recent 7.x image (currently [7.17.10](https://www.docker.elastic.co/r/elasticsearch/elasticsearch:7.17.10-amd64) *and* harden that image.

This means that the version of ES that we test against will be different from the version running on the cluster.

Spike to see if the tests even pass on 7.17.10. The easiest way to do that is to bump the image references (there are two) on a branch and push that branch to GitHub. GitHub pulls the upstream images, not the ones we mirror in ECR. Alternatively, one could set azul_docker_registry to `''` and run `make test`.

If that works, we need to create a new `Dockerfile` with, in essence 

```
FROM docker.elastic.co/elasticsearch/elasticsearch:7.17.10
RUN sudo apt-get update && sudo apt-get upgrade
```

and a Makefile that can build and push that image to Dockerhub (remember that that image needs to be multiarch).

- [ ] ~Security design review completed; the Resolution of this issue does **not**~ …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [ ] ~… introduce a new software dependency like~ (https://github.com/DataBiosphere/azul/issues/5292#issuecomment-1581656893)
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: #5292

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Deployed `dev.shared`
- [x] Check that the ECR repository `docker.elastic.co/elasticsearch/elasticsearch` was deleted in `dev`, file issue if it wasn't, don't delete the repository
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvildev.shared`
- [x] Check that the ECR repository `docker.elastic.co/elasticsearch/elasticsearch` was deleted in `anvildev`, file issue if it wasn't, don't delete the repository
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvilprod.shared`
- [x] Check that the ECR repository `docker.elastic.co/elasticsearch/elasticsearch` was deleted in `anvilprod`, file issue if it wasn't, don't delete the repository
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added to `prod` promotion PR:
> - [ ] Deployed `prod.shared`
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
What happened to peer review?

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Conflicts.

Currently, the ES docker image and the python package in `requirements.txt` are both on version 7.10.1. I'm surprised that this PR updates the former, but not the latter. Note that pip doesn't support elasicsearch==7.17.10 (it jumps from 7.17.9 to 8.0.0a1).

**Review by nadove-ucsc — DISMISSED** [review]

File: test/azul_test_case.py [hunk 1]
```diff
@@ -135,6 +135,13 @@ class RE(str):
                 ),
 
                 'Call to deprecated function (or staticmethod) patch_source_cache',
+
+                # FIXME: Remove after upgrading to a v8 Elasticsearch client
+                #        https://github.com/elastic/elasticsearch-py/issues/1698
```
Comment (hannes-ucsc) [inline test/azul_test_case.py]: FIXMEs have to refer to Azul issues, and those issues can't already be closed.

Do these deprecation warnings show up in a deployed Lambda as well? If so, they will probably be very prevalent, and therefore cause lots of noise. Please raise this in PL.

File: src/azul/es.py [hunk 2]
```diff
@@ -59,10 +59,13 @@ def _create_client(cls, host, port, timeout):
             aws_auth = CachedBotoAWSRequestsAuth(aws_host=host,
                                                  aws_region=aws.region_name,
                                                  aws_service='es')
-            return Elasticsearch(http_auth=aws_auth,
-                                 use_ssl=True,
-                                 verify_certs=True,
-                                 connection_class=RequestsHttpConnection, **common_params)
+            client = Elasticsearch(http_auth=aws_auth, use_ssl=True, verify_certs=True,
```
Comment (hannes-ucsc) [inline src/azul/es.py]: F

File: src/azul/__init__.py [hunk 3]
```diff
@@ -1439,7 +1439,7 @@ def docker_registry(self) -> str:
     # the `shared` TF component.
 
     docker_images = [
-        'docker.elastic.co/elasticsearch/elasticsearch:7.10.1',
+        'docker.elastic.co/elasticsearch/elasticsearch:7.17.10',
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: This needs to refer to our hardened image, obviously.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/es_test_case.py [hunk 4]
```diff
@@ -36,7 +36,8 @@ class ElasticsearchTestCase(DockerContainerTestCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
-        es_endpoint = cls._create_container('docker.elastic.co/elasticsearch/elasticsearch:7.10.1',
+        # TODO remove drop commit that put the non-hardened ES image here
+        es_endpoint = cls._create_container('docker.elastic.co/elasticsearch/elasticsearch:7.17.10',
```
Comment (hannes-ucsc) [inline test/es_test_case.py]: I bumped the internal version to 2 when merging your PR in the other repo. This is the image we should be using: https://hub.docker.com/layers/ucscgi/azul-elasticsearch/7.17.10-2/images/sha256-d787b3432b5722a833733047dad6009e493fa4fabc024d3bc7e6e5e54902bbb9?context=repo

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I don't think this should be labeled `partial`.

**Comment by dsotirho-ucsc** [comment]
Notes:
- ~The [ucscgi/azul-elasticsearch](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch/general) Docker repository is private.~
- ~Since GitHub can't pull the hardened ES image from the private Docker repository or from AWS ECR, this PR includes a `drop!` commit which replaces the private hardened ES image with the non-hardened ES image from elasticsearch.~

The Docker repository is now public, and the `drop!` commit has been dropped.

**Review by hannes-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by dsotirho-ucsc** [comment]
Previous approval was auto-dismissed due to a rebase on `develop` and resolving a merge conflict in `UPGRADING.rst`.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**